src/values/ArrayValue.js

```diff
@@ -63,13 +63,30 @@ function evaluatePossibleNestedOptimizedFunctionsAndStoreEffects(
       }
       throw e;
     }
+
+    // This is an incremental step from this list aimed to resolve a particular issue: #2452
+    //
+    // Assumptions:
+    // 1. We are here because the array op is pure, havocing of bindings is not needed.
+    // 2. The array op is only used once. To be enforced: #2448
+    // 3. Aliasing effects will lead to a fatal error. To be enforced: #2449
+    //
+    // Transitive materialization is needed to unblock this issue: #2405
+    //
+    // The bindings themselves do not have to materialize, since the values in them
+    // are used to optimize the nested optimized function. We compute the set of
+    // objects that are transitively reachable from read bindings and materialize them.
+
+    Materialize.materializeObjectsTransitive(realm, func);
```

Comment:
What about materializing the `abstractArrayValue`?

Reply:
Have an invariant that `abstractArrayValue` has already leaked?

Reply:
Yes, the index properties of widened arrays are not backed by locations, so no materialization is needed. I'll add an invariant that fails if this every changes.